### PR TITLE
possibility to change name in lobby

### LINKED ISSUE #2196: Name change
**Describe your proposal**
Only a small less important issue:
In OH3 it is possible to change your name in singleplayer player-selection. VCMI doesn't allow this.

Peek 2023-05-28 15-52

**Additional context**
Same functionality as OH3 should be supported

### PR DESCRIPTION
**Pull request by Laserlicht**
fixes #2196

### CODE REVIEW

File: server/CVCMIServer.cpp [hunk 1]
```diff
@@ -835,6 +835,19 @@ void CVCMIServer::setPlayer(PlayerColor clickedColor)
 	}
 }
 
+void CVCMIServer::setPlayerName(PlayerColor color, std::string name)
+{
+	PlayerSettings & player = si->playerInfos[color];
```
Comment (Nordsoft91) [inline server/CVCMIServer.cpp:840]:
```suggestion
	PlayerSettings & player = si->playerInfos.at(color);
```

Also, maybe test if color is valid player (CANNOT_DETERMINE is default value in cpack)

File: server/CVCMIServer.cpp [hunk 2]
```diff
@@ -835,6 +835,19 @@ void CVCMIServer::setPlayer(PlayerColor clickedColor)
...
+
+	if(player.isControlledByHuman())
+	{
+		int nameID = *(player.connectedPlayerIDs.begin()); //if not AI - set appropiate ID
```
Comment (Nordsoft91) [inline server/CVCMIServer.cpp:844]: what if player disconnected from lobby? make sense to check for empty?

**Comment by Nordsoft91** [comment]
3 bugs during testing:
 - hotseat: start changing name of player1 -> click on name field player2 -> change name of player2 -> press enter
 result: only player2 name is updated
 - (more generic) player name is updated only if return key pressed. Change name without pressing return key -> start game
 result: name is not updated
 hint: maybe update player name on event of text field change (see chess timers fields as an example)
- multiplayer: only host player can change the name of themselves. Name for joined players cannot be changed neither by host nor by client player

**Comment by Laserlicht** [comment]
>     * hotseat: start changing name of player1 -> click on name field player2 -> change name of player2 -> press enter
>       result: only player2 name is updated

fixed

>     * (more generic) player name is updated only if return key pressed. Change name without pressing return key -> start game
>       result: name is not updated
>       hint: maybe update player name on event of text field change (see chess timers fields as an example)

Yes. Tested this first. But it was not good when every keypress is send to server. Another problem is that after name change the lobby is updated (and the input field is also updated and lose the focus). I added a change if player clicks everywhere (other than the input field).
 
>     * multiplayer: only host player can change the name of themselves. Name for joined players cannot be changed neither by host nor by client player

Yes I disabled it (see `CSH->isGuest()`). Because the UI is locked (It's not possible to do enter and also not possible to send name change request to server). And honestly in Multiplayer it does make really sense because you asked for your name directly before the lobby. In Singleplayer you can currently change the name only with workarounds.

But if anyone find a solution for this in Multiplayer it's no problem to change it.

**Comment by Nordsoft91** [comment]
> And honestly in Multiplayer it does make really sense


Then maybe two options would be better than current behavior:
1. Allow host to change names for all guests
2. Disable name changing for host in MP games

**Comment by Laserlicht** [comment]
used second approach

**Comment by Nordsoft91** [comment]
some wierd bug:
- start hotseat with 2 players
- rename player1
- click on player2 name
- start typing
- observe hotkeys pressed and focus lost

**Comment by Laserlicht** [comment]
Yes. I know this. But it's a minor thing I think. I was not able to solve it.

Problem here is:
- rename player 1
- with clicking on player 2 name change on player 1 is accepted and send to server
- rename of player 2 is activated
- (few milliseconds later) server send new player names back -> this deactivate all text boxes
- text input isn't on textbox anymore -> hotkeys working again

**Review by Nordsoft91 — DISMISSED** [review]
Change breaks chess-timers setup functionality:
- enable vcmi-extras
- in options tab, press timer selection button
- select chess timer
- try to change timer limits

**Comment by Laserlicht** [comment]
@Nordsoft91 fixed; thanks!

File: client/CServerHandler.h [hunk 3]
```diff
@@ -67,6 +67,7 @@ class IServerAPI
 	virtual void setCampaignBonus(int bonusId) const = 0;
 	virtual void setMapInfo(std::shared_ptr<CMapInfo> to, std::shared_ptr<CMapGenOptions> mapGenOpts = {}) const = 0;
 	virtual void setPlayer(PlayerColor color) const = 0;
+	virtual void setPlayerName(PlayerColor color, std::string name) const = 0;
```
Comment (Nordsoft91) [inline client/CServerHandler.h:70]:
```suggestion
	virtual void setPlayerName(PlayerColor color, const std::string & name) const = 0;
```

File: client/CServerHandler.h [hunk 4]
```diff
@@ -152,6 +153,7 @@ class CServerHandler : public IServerAPI, public LobbyInfo
 	void setCampaignBonus(int bonusId) const override;
 	void setMapInfo(std::shared_ptr<CMapInfo> to, std::shared_ptr<CMapGenOptions> mapGenOpts = {}) const override;
 	void setPlayer(PlayerColor color) const override;
+	void setPlayerName(PlayerColor color, std::string name) const override;
```
Comment (Nordsoft91) [inline client/CServerHandler.h:156]:
```suggestion
	void setPlayerName(PlayerColor color, const std::string & name) const override;
```

File: client/lobby/OptionsTab.h [hunk 5]
```diff
@@ -186,8 +188,14 @@ class OptionsTab : public InterfaceObjectConfigurable
...
 		void hideUnavailableButtons();
+		bool captureThisKey(EShortcut key) override;
+		void keyPressed(EShortcut key) override;
+		void clickReleased(const Point & cursorPosition) override;
+		bool receiveEvent(const Point & position, int eventType) const override;
 
 	private:
 		const OptionsTab & parentTab;
+
+		void changeName();
```
Comment (Nordsoft91) [inline client/lobby/OptionsTab.h:199]:
```suggestion
		void updateName();
```

File: client/lobby/OptionsTab.cpp [hunk 6]
```diff
@@ -1115,6 +1123,43 @@ OptionsTab::PlayerOptionsEntry::PlayerOptionsEntry(const PlayerSettings & S, con
...
+	return eventType == AEventsReceiver::LCLICK; // capture all left clicks (not only within control)
+}
+
+void OptionsTab::PlayerOptionsEntry::clickReleased(const Point & cursorPosition)
+{
+	if(labelPlayerNameEdit && !labelPlayerNameEdit->pos.isInside(cursorPosition))
+		changeName();
+}
+
+void OptionsTab::PlayerOptionsEntry::changeName() {
```
Comment (Nordsoft91) [inline client/lobby/OptionsTab.cpp:1148]:
```suggestion
void OptionsTab::PlayerOptionsEntry::updateName() {
```

File: server/CVCMIServer.cpp [hunk 7]
```diff
@@ -835,6 +835,25 @@ void CVCMIServer::setPlayer(PlayerColor clickedColor)
...
+{
+	if(color == PlayerColor::CANNOT_DETERMINE)
+		return;
+
+	PlayerSettings & player = si->playerInfos.at(color);
+
+	if(!player.isControlledByHuman())
+		return;
+
+	if(!player.connectedPlayerIDs.size())
```
Comment (Nordsoft91) [inline server/CVCMIServer.cpp:848]:
```suggestion
	if(player.connectedPlayerIDs.empty())
```

**Review by Nordsoft91 — APPROVED** [review]

**Pull request merged**